File: PlantBasedPizza.Infra.Constructs/EventQueue.cs
using Amazon.CDK.AWS.EC2;
using Amazon.CDK.AWS.ECR;
using Amazon.CDK.AWS.ECS;
using Amazon.CDK.AWS.ElasticLoadBalancingV2;
using Amazon.CDK.AWS.Events;
using Amazon.CDK.AWS.Events.Targets;
using Amazon.CDK.AWS.IAM;
using Amazon.CDK.AWS.SQS;
using Amazon.CDK.AWS.SSM;
using Constructs;
using EventBus = Amazon.CDK.AWS.Events.EventBus;

namespace PlantBasedPizza.Infra.Constructs;

public record EventQueueProps(IEventBus Bus, string QueueName, string Environment, string EventSource, string DetailType);

public class EventQueue : Construct
{
    public IQueue Queue { get; private set; }
    
    public EventQueue(Construct scope, string id, EventQueueProps props) : base(scope, id)
    {
        var eventSource = props.EventSource;
        
        if (!props.EventSource.EndsWith("/"))
        {
            eventSource += "/";
        }
        
        this.Queue = new Queue(this, $"{props.QueueName}-{props.Environment}", new QueueProps()
        {
            QueueName = $"{props.QueueName}-{props.Environment}"
        });

        var rule = new Rule(this, $"{props.QueueName}Rule", new RuleProps()
        {
            EventBus = props.Bus
        });
        rule.AddEventPattern(new EventPattern()
        {
            Source = [eventSource],
            DetailType = [props.DetailType]
        });
        rule.AddTarget(new SqsQueue(this.Queue));
    }
}

File: src/PlantBasedPizza.Recipes.Java/infra/src/main/java/com/recipes/infrastructure/EventQueue.java
package com.recipes.infrastructure;

import org.jetbrains.annotations.NotNull;
import software.amazon.awscdk.Tags;
import software.amazon.awscdk.services.events.EventPattern;
import software.amazon.awscdk.services.events.Rule;
import software.amazon.awscdk.services.events.RuleProps;
import software.amazon.awscdk.services.events.targets.SqsQueue;
import software.amazon.awscdk.services.sqs.DeadLetterQueue;
import software.amazon.awscdk.services.sqs.IQueue;
import software.amazon.awscdk.services.sqs.Queue;
import software.amazon.awscdk.services.sqs.QueueProps;
import software.constructs.Construct;

import java.util.ArrayList;
import java.util.List;

public class EventQueue extends Construct {
    private final Queue deadLetterQueue;
    private final Queue queue;
    
    public EventQueue(@NotNull Construct scope, @NotNull String id, @NotNull EventQueueProps props) {
        super(scope, id);
        String eventSource = props.getEventSource();
        
        if (!eventSource.endsWith("/"))
        {
            eventSource = eventSource + "/";
        }
        
        String deadLetterQueueName = String.format("%sDLQ-%s", props.getQueueName(), props.getSharedProps().getEnvironment());
        
        this.deadLetterQueue = new Queue(this, deadLetterQueueName, 
                QueueProps.builder()
                        .queueName(deadLetterQueueName)
                        .build());

        String queueName = String.format("%s-%s", props.getQueueName(), props.getSharedProps().getEnvironment());
        
        this.queue = new Queue(this, queueName,
                QueueProps.builder()
                        .queueName(deadLetterQueueName)
                        .deadLetterQueue(DeadLetterQueue.builder()
                                .queue(this.deadLetterQueue)
                                .maxReceiveCount(3).build())
                        .build());

        Tags.of(this.queue).add("service", props.getSharedProps().getServiceName());
        Tags.of(this.deadLetterQueue).add("service", props.getSharedProps().getServiceName());

        Rule rule = new Rule(this, String.format("%sRule", props.getQueueName()),
                RuleProps.builder()
                        .eventBus(props.getBus())
                        .build());
        
        List<String> sources = new ArrayList<>();
        sources.add(eventSource);

        List<String> detailTypes = new ArrayList<>();
        detailTypes.add(props.getDetailType());
        
        EventPattern pattern = EventPattern
                .builder()
                .source(sources)
                .detailType(detailTypes)
                        .build();
        rule.addEventPattern(pattern);
        
        rule.addTarget(new SqsQueue(this.queue));
    }
    
    public IQueue getQueue() {
        return this.queue;
    }

    public IQueue getDeadLetterQueue() {
        return this.deadLetterQueue;
    }
}
